File: python/06_reports.py
"""
Cynco API Quickstart -- 06: Reports

Demonstrates how to pull accounting reports from the Cynco API.
Fetches a trial balance and formats it as a readable table.

Run: python 06_reports.py
"""

import os
import sys
from datetime import date

import requests

API_KEY: str = os.environ.get("CYNCO_API_KEY", "")
BASE_URL: str = "https://app.cynco.io/api/v1"

if not API_KEY:
    print("Missing CYNCO_API_KEY environment variable.")
    print("Set it with: export CYNCO_API_KEY='cak_your_api_key_here'")
    sys.exit(1)

headers: dict[str, str] = {
    "Authorization": f"Bearer {API_KEY}",
}


def format_amount(cents: int, currency: str) -> str:
    """
    Format a monetary value from cents to a readable string.
    Cynco API returns all monetary values in the smallest currency unit.
    """
    value = abs(cents) / 100
    return f"{currency} {value:>12,.2f}"


def main() -> None:
    """Fetch and display a trial balance report."""

    # Fetch the trial balance as of today.
    # You can pass any date for historical reporting.
    today: str = date.today().isoformat()

    response = requests.get(
        f"{BASE_URL}/reports/trial-balance",
        headers=headers,
        params={"asOf": today},
        timeout=30,
    )

    if response.status_code == 401:
        print("Invalid API key. Check your CYNCO_API_KEY environment variable.")
        sys.exit(1)

    response.raise_for_status()

    body: dict = response.json()
    report: dict = body["data"]
    currency: str = report["currency"]
    accounts: list[dict] = report["accounts"]
    totals: dict = report["totals"]

    # Print the report header.
    width: int = 78
    print(f"\nTrial Balance as of {today}")
    print("=" * width)

    # Column headers.
    print(
        f"{'Code':<8} {'Account':<30} {'Debit':>18} {'Credit':>18}"
    )
    print("-" * width)

    # Print each account line.
    for account in accounts:
        debit_str = format_amount(account["debit"], currency) if account["debit"] > 0 else ""
        credit_str = format_amount(account["credit"], currency) if account["credit"] > 0 else ""

        print(
            f"{account['code']:<8} "
            f"{account['name']:<30} "
            f"{debit_str:>18} "
            f"{credit_str:>18}"
        )

    # Print totals.
    print("-" * width)

    total_debit = format_amount(totals["debit"], currency)
    total_credit = format_amount(totals["credit"], currency)

    print(
        f"{'':8} {'TOTAL':<30} {total_debit:>18} {total_credit:>18}"
    )
    print("=" * width)

    # Verify the trial balance.
    if totals["debit"] == totals["credit"]:
        print("\nTrial balance is in balance.")
    else:
        difference = abs(totals["debit"] - totals["credit"])
        diff_str = format_amount(difference, currency)
        print(f"\nWARNING: Trial balance is out of balance by {diff_str}")

    # Other available reports:
    #
    # Profit & Loss:
    #   GET /reports/profit-loss?from=2026-01-01&to=2026-03-31
    #
    # Balance Sheet:
    #   GET /reports/balance-sheet?asOf=2026-03-31
    #
    # Aged Receivables:
    #   GET /reports/aged-receivables?asOf=2026-03-31


if __name__ == "__main__":
    main()

File: typescript/src/06-reports.ts
/**
 * Cynco API Quickstart -- 06: Reports
 *
 * Demonstrates how to pull accounting reports from the Cynco API.
 * This example fetches a trial balance and formats it for display.
 *
 * Run: npm run example:reports
 */

import Cynco from "@cynco/sdk";

const cynco = new Cynco({
  apiKey: process.env.CYNCO_API_KEY,
});

/**
 * Format a monetary value from cents to a readable string.
 * Cynco API returns all monetary values in the smallest currency unit.
 */
function formatAmount(cents: number, currency: string): string {
  const value = Math.abs(cents) / 100;
  const formatted = value.toLocaleString("en-MY", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
  return `${currency} ${formatted}`;
}

/**
 * Pad a string to a fixed width for table formatting.
 */
function padRight(str: string, width: number): string {
  return str.length >= width ? str : str + " ".repeat(width - str.length);
}

function padLeft(str: string, width: number): string {
  return str.length >= width ? str : " ".repeat(width - str.length) + str;
}

async function main(): Promise<void> {
  // Fetch the trial balance as of today.
  // You can pass any date -- the API supports historical reporting.
  const today = new Date().toISOString().split("T")[0];

  console.log(`Trial Balance as of ${today}`);
  console.log("=".repeat(78));

  const report = await cynco.reports.trialBalance({
    asOf: today,
  });

  // Print the header.
  console.log(
    `${padRight("Code", 8)} ${padRight("Account", 30)} ${padLeft("Debit", 18)} ${padLeft("Credit", 18)}`
  );
  console.log("-".repeat(78));

  // Print each account line.
  for (const account of report.accounts) {
    const debit =
      account.debit > 0
        ? formatAmount(account.debit, report.currency)
        : "";
    const credit =
      account.credit > 0
        ? formatAmount(account.credit, report.currency)
        : "";

    console.log(
      `${padRight(account.code, 8)} ${padRight(account.name, 30)} ${padLeft(debit, 18)} ${padLeft(credit, 18)}`
    );
  }

  // Print totals.
  console.log("-".repeat(78));

  const totalDebit = formatAmount(report.totals.debit, report.currency);
  const totalCredit = formatAmount(report.totals.credit, report.currency);

  console.log(
    `${padRight("", 8)} ${padRight("TOTAL", 30)} ${padLeft(totalDebit, 18)} ${padLeft(totalCredit, 18)}`
  );
  console.log("=".repeat(78));

  // Verify the trial balance balances (debits = credits).
  if (report.totals.debit === report.totals.credit) {
    console.log("\nTrial balance is in balance.");
  } else {
    const difference = Math.abs(report.totals.debit - report.totals.credit);
    console.log(
      `\nWARNING: Trial balance is out of balance by ${formatAmount(difference, report.currency)}`
    );
  }

  // You can also fetch other reports with similar patterns:
  //
  // Profit & Loss:
  //   const pnl = await cynco.reports.profitAndLoss({
  //     from: "2026-01-01",
  //     to: "2026-03-31",
  //   });
  //
  // Balance Sheet:
  //   const bs = await cynco.reports.balanceSheet({
  //     asOf: "2026-03-31",
  //   });
  //
  // Aged Receivables:
  //   const ar = await cynco.reports.agedReceivables({
  //     asOf: "2026-03-31",
  //   });
}

main().catch((error: unknown) => {
  if (error instanceof Cynco.AuthenticationError) {
    console.error("Invalid API key. Check your CYNCO_API_KEY environment variable.");
  } else {
    console.error("Failed to generate report:", error);
  }
  process.exit(1);
});
